## Programacion Estructurada

In [6]:
import numpy as np
from scipy.optimize import minimize

# funcion
def f(params):
    x, y = params
    term1 = x**2 * (4 - 2.1 * x**2 + (x**4 / 3))
    term2 = x * y
    term3 = y**2 * (-4 + 4 * y**2)
    return term1 + term2 + term3

# punto inicial de busqueda 
punto_inicial = [0.1, -0.5]

resultado = minimize(f, punto_inicial)

print(f"min = {resultado.x[0]:.4f}, y = {resultado.x[1]:.4f}")
print(f"valor en el punto: {resultado.fun:.4f}")

min = 0.0898, y = -0.7127
valor en el punto: -1.0316


## Codigo del libro - algoritmo de enjambre de partículas

In [1]:
from random import random, randint

Función que se quiere minimizar

In [2]:
def funcion(x, y):
    sum1 = x**2 * (4 - 2.1 * x**2 + x**4 / 3.0)
    sum2 = x * y
    sum3 = y**2 * (-4 + 4 * y**2)
    return sum1 + sum2 + sum3

# Devuelve un número aleatorio dentro de un rango con 
# distribución uniforme (proporcionada por random)
def aleatorio(inf, sup):
    return random() * (sup - inf) + inf

Clase que representa una partícula individual y que facilita

In [3]:
class Particula:
    # Algunos atributos de clase (comunes a todas las partículas)
    # Parámetros para actualizar la velocidad
    inercia   = 1.4
    cognitiva = 2.0
    social    = 2.0
    
    # Límites del espacio de soluciones
    infx = -2.0
    supx = 2.0
    infy = -1.0
    supy = 1.0
    
    # Factor de ajuste de la velocidad inicial
    ajusteV = 100.0

    # Crea una partícula dentro de los límites indicados
    def __init__(self):
        self.x = aleatorio(Particula.infx, Particula.supx)
        self.y = aleatorio(Particula.infy, Particula.supy)
        
        self.vx = aleatorio(Particula.infx / Particula.ajusteV, 
                            Particula.supx / Particula.ajusteV)
        self.vy = aleatorio(Particula.infy / Particula.ajusteV, 
                            Particula.supy / Particula.ajusteV)
        
        self.xLoc = self.x
        self.yLoc = self.y
        self.valorLoc = funcion(self.x, self.y)

    # Actualiza la velocidad de la partícula
    def actualizaVelocidad(self, xGlob, yGlob):
        cogX = Particula.cognitiva * random() * (self.xLoc - self.x)
        socX = Particula.social * random() * (xGlob - self.x)
        self.vx = Particula.inercia * self.vx + cogX + socX
        
        cogY = Particula.cognitiva * random() * (self.yLoc - self.y)
        socY = Particula.social * random() * (yGlob - self.y)
        self.vy = Particula.inercia * self.vy + cogY + socY

    # Actualiza la posición de la partícula
    def actualizaPosicion(self):
        self.x = self.x + self.vx
        self.y = self.y + self.vy

        # Debe mantenerse dentro del espacio de soluciones
        self.x = max(self.x, Particula.infx)
        self.x = min(self.x, Particula.supx)
        self.y = max(self.y, Particula.infy)
        self.y = min(self.y, Particula.supy)

        # Si es inferior a la mejor, la adopta como mejor
        valor = funcion(self.x, self.y)
        if valor < self.valorLoc:
            self.xLoc = self.x
            self.yLoc = self.y
            self.valorLoc = valor

# Devuelve las coordenadas y el valor del mínimo obtenido.

In [4]:
def enjambreParticulas(particulas, iteraciones, reduccionInercia):
    
    # Registra la mejor posición global y su valor
    mejorParticula = min(particulas, key=lambda p: p.valorLoc)
    xGlob = mejorParticula.xLoc
    yGlob = mejorParticula.yLoc
    valorGlob = mejorParticula.valorLoc

    # Bucle principal de simulación
    for iter in range(iteraciones):
        # Actualiza la velocidad y posición de cada partícula
        for p in particulas:
            p.actualizaVelocidad(xGlob, yGlob)
            p.actualizaPosicion()

        # Hasta que no se han movido todas las partículas no se 
        # actualiza el mínimo global, para simular que todas se
        # mueven a la vez
        mejorParticula = min(particulas, key=lambda p: p.valorLoc)
        if mejorParticula.valorLoc < valorGlob:
            xGlob = mejorParticula.xLoc
            yGlob = mejorParticula.yLoc
            valorGlob = mejorParticula.valorLoc

        # Finalmente se reduce la inercia de las partículas
        Particula.inercia *= reduccionInercia

    return (xGlob, yGlob, valorGlob)

Parametros y Ejecucion

In [5]:
nParticulas = 10
iteraciones = 100
redInercia = 0.9

# Genera un conjunto inicial de partículas
particulas = [Particula() for i in range(nParticulas)]

# Ejecuta el algoritmo del enjambre de partículas
print(enjambreParticulas(particulas, iteraciones, redInercia))

(-0.08984201711583026, 0.7126564042717368, -1.0316284534898774)


## Comparacion 

- Prog. estructurada : (x, y) = (0.0898, -0.7127), f(x,y) = -1.0316
- Enjambre de partículas (libro) : (x, y) = (-0.0898, 0.7127)`, f(x,y) = -1.0316

Conclusion =  los dos métodos encuentran exactamente el mismo mínimo (pero uno siendo mas especifico que el otro), solo cambia los signos. 